### PR TITLE
[529] Build xtable with scala version(s)

### PR DESCRIPTION
**Pull request by rangareddy**
## *Important Read*

This pull request aims to address issues [#529](https://github.com/apache/incubator-xtable/issues/529)

## What is the purpose of the pull request

To support building the xtable with multiple scala versions i.e 2.12 and 2.13. 

## Brief change log

* Added scala12 and scala13 version properties in pom.xml file.

```xml
<scala12.version>2.12.15</scala12.version>
<scala13.version>2.13.8</scala13.version>
```

* Added profiles to support scala12 and scala13

```xml
<!--Scala 2.12 Profile -->
<profile>
    <id>scala-2.12</id>
    <activation>
        <activeByDefault>true</activeByDefault>
    </activation>
    <properties>
        <scala.version>${scala12.version}</scala.version>
        <scala.version.prefix>2.12</scala.version.prefix>
    </properties>
    <build>
        <pluginManagement/>
    </build>
</profile>

<!--Scala 2.13 Profile -->
<!-- Once hudi supports scala 2.13 then enable following profile -->
<!--<profile>
    <id>scala-2.13</id>
    <activation>
        <activeByDefault>false</activeByDefault>
    </activation>
    <properties>
        <scala.version>${scala13.version}</scala.version>
        <scala.version.prefix>2.13</scala.version.prefix>
    </properties>
    <build>
        <pluginManagement>
            <plugins>
                <plugin>
                    <groupId>net.alchim31.maven</groupId>
                    <artifactId>scala-maven-plugin</artifactId>
                    <configuration>
                        <args>
                            <arg>-unchecked</arg>
                            <arg>-deprecation</arg>
                            <arg>-feature</arg>
                            <arg>-explaintypes</arg>
                            <arg>-target:jvm-1.8</arg>
                        </args>
                        <compilerPlugins/>
                    </configuration>
                </plugin>
            </plugins>
        </pluginManagement>
    </build>
</profile>-->
```

## Verify this pull request

Run `$ mvn clean install -DskipTests -U` locally.

**Note:** Hudi does not yet support Scala 2.13 jars, so the scala-2.13 profile is commented out. Once Hudi supports Scala 2.13, we can use the following command to build xtable with Scala 2.13.

```sh
mvn clean install -DskipTests -Pscala-2.13 
```

## References

* https://github.com/apache/spark/blob/branch-3.4/pom.xml#L3617
* https://github.com/apache/incubator-gluten/blob/main/pom.xml#L132

### CODE REVIEW

**Comment by vinishjail97** [comment]
Let's suffix the Scala version (2.12) to the jar name as suggested by @pjfanning 
https://github.com/apache/incubator-xtable/issues/529#issuecomment-2358231091

File: pom.xml [hunk 1]
```diff
@@ -68,8 +69,10 @@
         <maven-deploy-plugin.version>3.1.1</maven-deploy-plugin.version>
         <maven-release-plugin.version>2.5.3</maven-release-plugin.version>
         <parquet.version>1.12.2</parquet.version>
-        <scala.version>2.12.15</scala.version>
-        <scala.version.prefix>2.12</scala.version.prefix>
+        <scala12.version>2.12.15</scala12.version>
+        <scala13.version>2.13.8</scala13.version>
```
Comment (pjfanning) [inline pom.xml:73]: why not use Scala 2.12.20 and 2.13.14 - the latest versions of these patch lines?

Comment (rangareddy) [reply pom.xml:73]: Based on the Spark version, we need to use a supported Scala version. Specifically, **Spark 3.4** version supports **Scala 2.12.17** for **Scala 2.12** and **Scala 2.13.8** for **Scala 2.13**.

Whenever I address all review comments, I will update the Scala 2.12 version to 2.12.17.

**Reference:**

1. https://github.com/apache/spark/blob/branch-3.4/pom.xml#L3624

Comment (pjfanning) [reply pom.xml:73]: You should be able to use the latest patch level Scala compiler releases. This is what the Scala team recommend.

File: Dockerfile [hunk 2]
```diff
@@ -23,7 +23,8 @@ WORKDIR /build
 COPY ./ ./
 RUN --mount=type=cache,target=/root/.m2 \
     MAVEN_OPTS=-Dorg.slf4j.simpleLogger.defaultLogLevel=warn mvn  -B  package -DskipTests 
-RUN mv xtable-utilities/target/xtable-utilities-$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-bundled.jar target/app.jar
+RUN mv xtable-utilities_2.12-$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-bundled.jar target/app.jar
+#RUN mv xtable-utilities/target/$(mvn help:evaluate -Dexpression=project.build.finalName -q -DforceStdout)-bundled.jar target/app.jar
```
Comment (vinishjail97) [inline Dockerfile:27]: We can remove the line that's commented.

**Review by vinishjail97 — COMMENTED** [review]
There are still some references for these jars across the repo. Can include the suffix in those places as well ? For example, website module just uses 0.2.0-SNAPSHOT without the scala suffix for utilities and core jars. 

https://github.com/search?q=repo%3Aapache%2Fincubator-xtable%20xtable-utilities-&type=code. 
 https://github.com/search?q=repo%3Aapache%2Fincubator-xtable+xtable-core-&type=code

File: Dockerfile (hunk 2 above)
Comment (rangareddy) [reply Dockerfile]: Done

**Comment by vinishjail97** [comment]
@pjfanning Can you review if things look okay ? This is the last release blocker for 0.2.0.

File: pom.xml [hunk 3]
```diff
@@ -867,6 +871,53 @@
...
+            <properties>
+                <scala.version>${scala12.version}</scala.version>
+                <scala.binary.version>2.12</scala.binary.version>
+            </properties>
+            <build>
+                <pluginManagement/>
+            </build>
+        </profile>
+
+        <!--Scala 2.13 Profile -->
```
Comment (vinishjail97) [inline pom.xml:889]: @pjfanning Do we leave this as commented out or should we clean it up ?

**Review by pjfanning — APPROVED** [review]
lgtm

**Review by vinishjail97 — APPROVED** [review]

File: Dockerfile [hunk 4]
```diff
@@ -23,7 +23,7 @@ WORKDIR /build
 COPY ./ ./
 RUN --mount=type=cache,target=/root/.m2 \
     MAVEN_OPTS=-Dorg.slf4j.simpleLogger.defaultLogLevel=warn mvn  -B  package -DskipTests 
-RUN mv xtable-utilities/target/xtable-utilities-$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-bundled.jar target/app.jar
+RUN mv xtable-utilities_2.12-$(mvn help:evaluate -Dexpression=project.version -q -DforceStdout)-bundled.jar target/app.jar
```
Comment (vinishjail97) [inline Dockerfile:26]: @rangareddy I ran this locally, should this be ? 
`xtable-utilities/target/xtable-utilities_2.12`

**Comment by vinishjail97** [comment]
@rangareddy Let's squash before merging.

**Pull request merged**